doc/rpld.8
.TH "rpld" "8" "2009-09-27" "ttyrpld" "tty logging daemon suite"
.SH "Name"
.PP
rpld \(em tty logging daemon
.SH Syntax
.PP
\fBrpld\fP [\fB\-IQsv\fP] [\fB\-D\fP \fIdevice\fP] [\fB\-O\fP \fIofmt\fP]
[\fB\-U\fP \fIuser\fP] [\fB\-c\fP \fIconfigfile\fP]
.SH Description
.PP
rpld is the user-space daemon that reads /dev/rpl and plexes the data to
different files depending on which tty they were logged. It also adds a
timestamp so that replaying can be done in real-time.
.SH Options
.PP
.TP
\fB\-D\fP \fIdevice\fP
Path to the rpl device, e.g. \fB/dev/rpl\fP.
.TP
\fB\-I\fP
Start the infod component if it is not automatically started when the
INFOD_START configuration variable is set.
.TP
\fB\-O\fP \fIofmt\fP
Overrides the hardcoded and configuration file values (processed so far) for
the log file naming scheme. You can use the printf-like placeholders explained
below. Subsequent \fB\-c\fP options may override this, if an OFMT variable is
found in further configuration files. Make sure the user can create files
according to \fIofmt\fP.
.TP
\fB\-Q\fP
All ttys start in deactivated mode (rather than activated). They can then be
activated when needed. Byte-counting is done in either case.
.TP
\fB\-U\fP \fIuser\fP
Drop all privilegues and change to \fIuser\fP after initialization (memory lock,
creating and opening RPL device). This can be either a username or a numeric
UID.
.TP
\fB\-c\fP \fIfile\fP
Load configuration variables from \fIfile\fP. The in-memory copies of the
variables are not changed if no such var name is found in the file. Unknown var
names in the configuration file are also ignored.
.TP
\fB\-s\fP
Print warnings and error messages to syslog rather than stderr.
.TP
\fB\-v\fP
Print statistics about recorded packets on stdout while rpld is running. This
option overrides \fB-s\fP.
.SH "Privilege separation"
.PP
rpld offers the possibility to change to another user's identity after the
initialization phase is complete. The default package (from ttyrpld.sf.net)
uses the daemon user. You can change this in the configuration file.
.SH "Logging"
.PP
\fBrpld\fP does not detach itself to help debugging, but using the
\fBstartproc\fP(8), \fBsetsid\fP(1) starter tools or bash's disown builtin can
help bringing it in the background, if you need to. The only time when the
daemon will output something is either during initialization or when there is
really trouble, like memory allocation failure. In either case, error messages
are rare and you would not need to worry about redirecting stderr.
.PP
You can run rpld with the \-v option to enable printing statistics on stdout.
Even if you do not have \-v specified, you can send rpld a SIGALRM signal to
make it print the current statistics. That of course only makes sense when
stdout is connected to something but /dev/null or /dev/zero. You can send it
multiple SIGALRMs, of course, but you should consider using \-v then, maybe.
The statistics are printed with the move-to-beginning-of-line character (\r),
which is not that suitable for logging, though.
.PP
Basically, every tty is monitored, but certain kinds are excluded, like the
master sides of BSD (major number 2 (Linux), 6 on BSD) and Unix98 ptys (major
number 128), because they are just a mirror of their slave sides with things
turned around and are rarely useful. (On BSD, they are not explicitly excluded
because they do not seem to generate any data.)
.PP
rpld will divert logging of a tty to a new file when the tty inode is opened
the next time and if the owner of it has changed. This will make logins on
/dev/ttyN go to the right file. Note that \fBsu\fP(8) does not change the
ownership, so both sessions (normal and su'ed) will go into the same file. (And
that is good, because it then logs by "real person" rather than login name.)
The byte count statistics are zeroed upon change detection.
.SH "Configuration file".
.PP
rpld starts with its hardcoded defaults, then tries to read
\fB/etc/rpld.conf\fP and finally \fBrpld.conf\fP in the directory where the
rpld binary is in.
.PP
Configuration files have a simple KEY=VALUE syntax. Empty lines, lines
beginning with a hash mark (#) or unrecognized keys are ignored.
.TP
\fBDEVICE=\fP\fIdevice\fP
List of possible rpl devices, separated by space. The default is "/dev/misc/rpl
/dev/rpl".
.TP
\fBOFMT=\fP\fIformat\fP
OFMT combines the directory to write the tty recordings into (relative to the
working directory of rpld) and the name of the log files. Note that the working
directory of rpld depends on where it was started from, e.g. with startproc,
the CWD is mostly the root path (/). It is wise to use absolute paths here. The
available printf-style tags are below.\(em Directories will be created as
needed (if permission allows). Thus, you can have
/var/log/rpl/%(USER)/%(DATE).%(TIME).%(TTY) it will automatically
create the user directory, provided that /var/log/rpl is writable for
the user running rpld.
.TP
\fBUSER=\fP\fIuser\fP
User to change to after all initialization. Make sure that the user can create
files according to OFMT.
.TP
\fBSTART_INFOD=\fP\fIyes\fP|\fIno\fP
Starts the infod component which makes statistics available to \fBrplctl\fP(8).
.TP
\fBINFOD_PROT=\fP\fIpath\fP
Specify the path of the socket which infod provides for clients. The default
value is /var/run/.rplinfod_socket.
.SS "OFMT tags"
.PP
.TP
\fB%(DATE)\fP, \fB%(TIME)\fP
Date (in YYYYMMDD) and time (in HHMMSS format) when tty was opened (usually
when someone logs in or an xterm was opened).
.TP
\fB%(TTY)\fP
Terminal (line) the user logged on. The string is taken from a string database.
Typical strings are tty* for virtual consoles, pts-* for pseudo-terminals,
ttyS* for serial lines. If there is no matching string entry, the device number
is used in square brackets, e.g. [240:0].
.TP
\fB%(USER)\fP
User who owned the tty when it was opened. If the device could not be
stat(2)'ed, %u will be substituted by "NONE". If the UID could not be
translated into a name, the UID is spit out instead.
.SH "See also"
.PP
\fBttyreplay\fP(1), \fBrplctl\fP(8)


doc/changelog.txt

Notes
=====

You do not need to repatch your kernel or recompile rpldev if they have not
changed significantly between ttyrpld releases.

│                │last time kpatch│                            │
│Operating system│(rpldhk) changed│last time rpldev changed:   │
├────────────────┼────────────────┼────────────────────────────┤
│Linux           │2.50            │2.60                        │
│FreeBSD         │2.60            │2.60                        │
│OpenBSD         │-               │2.60                        │
│NetBSD          │2.50            │2.60                        │


HEAD
====
Changes:
- build: update to libHX 3.12 API
- rpldev: Linux 3.3 patch


v2.60 (2009-10-09)
==================
Changes:
- doc: move changelog to text/plain
- rpldev: NetBSD 5.0.1 support
- rpldev: OpenBSD 4.5 support
- rpldev: Linux 2.6.31 support
- rpldev: FreeBSD 8.0 support
- packet: increase the "size" field to 32 bits
- packet: combine "magic" and "event" fields, and move to front
Enhancements:
- rpld: filter certain interruptive escape


v2.52 (2008-09-07)
==================
Changes:
- Updated kernel components for Linux 2.6.27
- Updated userspace code for libHX 1.25


v2.51 (2008-05-15)
==================
Changes:
- Updated rpldhk and rpldev for Linux 2.6.25, OpenBSD 4.3, FreeBSD 7.0,
  NetBSD 4.0


v2.50 (2008-01-01)
==================
Changes:
- changed the packet format to be compatible between 32-bit and 64-bit
Enhancements:
- added rplcvt utility for converting old on-disk rpld files
- explicitly blacklist /dev/stdout, /dev/stderr and /dev/stderr - these have
  been popping up as tty names because there is such a symlink in Linux's /
  dev. (But we need symlink processing for Solaris, eww.)


v2.19 (2007-09-17)
==================
Fixes:
- fix a file descriptor leak in rpld (thanks J.Eller)
- fix a symlink recursion when walking /dev (J.Eller)
- raise maximum number of open files (J.Eller)
- handle permission restrictions on /dev nodes
- fix directory name of Persian translation
Changes:
- The kernel interface has been reduced (hooks removed). The 2.19 userspace
  is still compatible (throws warnings though) with 2.18 kernel components
  however. The ttyrpld 2.19 kernel patch is for Linux 2.6.22.
- moved to autotools
- build with new libHX 1.10
Enhancements:
- an experimental Solaris kernel driver has been added - tested with Solaris
  11.snv only, but may work with older ones.


v2.18 (2006-07-01)
==================
Fixes:
- Fixed OpenBSD 3.9 kpatch
- Readded workaround for OpenBSD/NetBSD EINTR signal handling that got lost
  in 2.17


v2.17 (2006-06-19)
==================
Changes:
- Support for Linux 2.6.17, FreeBSD 6.1, OpenBSD 3.9
Fixes:
- added missing -fprofile-arcs -ftest-coverage to LDFLAGS for PROF=1


v2.16 (2006-04-01)
==================
Changes:
- OpenBSD part is again just a patch since modules are said to be unsupported
  on non-i386
- Due to popular request, the docs now come with a black-on-white ("paper
  color") color scheme and no fixed font size. The old gray-on-black
  ("console color") is available as an alternate stylesheet and can be used
  in Firefox using View > Page Style > Black background.
Enhancements:
- Added polling functions to the BSD rpldevs. This fixes the high CPU usage
  (which is due to the implementation of pthreads on OpenBSD) observed with
  rpld/OpenBSD.
- Added the EVT_MAGIC pseudo event type
- Added share/ttyrpld.magic for the file(1) utility


v2.15 (2006-03-11)
==================
General:
- added support for NetBSD 3.0
- shifts for majors>255 and minors>255 were broken in include/dev.h

kpatch:
- [All] fixed some SMP race conditions
- [OpenBSD] fixed: rpl_deinit() was not called in ttyclose()

rpldev:
- [All] rename functions to make function more clear (mv_to_user ->
  circular_get, mv_buffer -> circular_put_packet, mv_buffer2 -> circular_put)
- [All] make avail_R() and avail_W() return size_t instead of ssize_t,
  addressing some potential overflow problems
- [Linux] document circulary calculation of avail_*(), please report errors
- [FreeBSD] merged freebsd-*/rpldev.c into one file
- [FreeBSD/OpenBSD] removed SKIP_PTM, since master ptys do not generate any
  noise (that's a kpatch feature)
- [OpenBSD] rpldev is now a module (rpldev_mod.o)
- [OpenBSD] loads at new device number (228,0), please re-mknod your /dev/rpl
  node


v2.12 (2006-01-21)
==================
- fixed: FreeBSD 6.0 hands krpl_close() a NULL pointer in some cases; handle
  it properly
- some doc updates
- GNUmakefile: "all" target does not include "doc" anymore, because BSDs
  (standard installation) often ship without php, w3m and perl.
- added the '[' and ']' keys to ttyreplay (lower/raise speed by 0.10)


v2.11 (2006-01-06)
==================
Added support back in for the latest releases of the previous generation of
kernels. For Linux, that's 2.4.something (USE DISCOURAGED!), and for FreeBSD
that's 5.3.


v2.10 (2005-12-04)
===================
Support for OpenBSD 3.8 and FreeBSD 6.0 was added; permission denied, syslog
spamming, cpu hog, ownership and symlink recursion bugs were fixed. Docs were
updated to reflect advances since the last seven releases.

- Support for OpenBSD 3.8 and FreeBSD 6.0
- Documentation has been updated to reflect advances since the last seven
  releases.
- user/replay.c: fixed a color bug when -t was active (spotted whilst under
  FreeBSD)
- user/*.c: use usleep(10000) instead of sched_yield() to reduce idle cpu hog
- user/rpld.c: umask was set wrong, so 1.) logging did not take place 2.)
  syslog was spammed with permission denied warnings
- user/rpld.c: added a rate limiter for syslog messages
- user/rpld.c: find_devnode_dive() forgot to check for symlinks so that we
  dived into /dev/fd, which is not wanted, because /dev/fd/* itself includes
  /dev again, which lead to unwanted results.
- user/rpld.c: fixed a filedescriptor leak in fill_info()
- user/rpld.c: another try at getting ownerships right; rpld still had a case

doc/rpld.8.php
<?php include_once("_header.php"); ?>

<h1>rpld - tty logging daemon</h1>

<p><code>rpld <b>[</b>-D <i>dev</i><b>] [</b>-I<b>] [</b>-O <i>ofmt</i><b>]
[</b>-Q<b>] [</b>-U <i>user</i><b>] [</b>-c <i>configfile</i><b>] [</b>-s<b>]
[</b>-v<b>]</b></code></p>

<h1>Description</h1>

<p class="block"><code>rpld</code> is the user-space daemon that reads
<code>/dev/rpl</code> and plexes the data to different files depending on which
tty they were logged. It also adds a timestamp so that replaying can be done in
real-time.</p>

<h1>Options</h1>

<table border="1">
  <tr>
    <td class="t1" nowrap="nowrap"><code><b>-D</b> <i>dev</i></code></td>
    <td class="t1">Path to the rpl device, e.g. <code>/dev/rpl</code>.</td>
  </tr>
  <tr>
    <td class="t2" nowrap="nowrap"><code><b>-I</b></code></td>
    <td class="t2">Start the <i>infod</i> component if it is not automatically
      started when the <code>INFOD_START</code> configuration variable is
      set.</td>
  </tr>
  <tr>
    <td class="t1" nowrap="nowrap"><code><b>-O</b> <i>string</i></code></td>
    <td class="t1">Overrides the hardcoded and configuration file values
      (processed so far) for the log file naming scheme. You can use the
      printf-like placeholders explained below.  Subsequent <code>-c</code> may
      override this, if an <code>OFMT</code> variable is found in further
      configuration files. Make sure the user can create files according to
      <code>-O</code>/<code>OFMT</code>.</td>
  </tr>
  <tr>
    <td class="t2" nowrap="nowrap"><code><b>-Q</b></code></td>
    <td class="t2">All ttys start in deactivated mode (rather than activated).
      They can then be activated when needed.  Byte-counting is done in either
      case.</td>
  </tr>
  <tr>
    <td class="t1" nowrap="nowrap"><code><b>-U</b> <i>user</i></code></td>
    <td class="t1">Drop all privilegues and change to <i>user</i> after
      initialization (memory lock, creating and opening RPL device). This can
      be either a username or a numeric UID.</td>
  </tr>
  <tr>
    <td class="t2" nowrap="nowrap"><code><b>-c</b> <i>file</i></code></td>
    <td class="t2">Load configuration variables from <i>file</i>. The in-memory
      copies of the variables are not changed if no such var name is found in
      the file. Unknown var names in the configuration file are also
      ignored.</td>
  </tr>
  <tr>
    <td class="t1" nowrap="nowrap"><code><b>-s</b></code></td>
    <td class="t1">Print warnings and error messages to <i>syslog</i> rather
      than <code>stderr</code>.</td>
  </tr>
  <tr>
    <td class="t2" nowrap="nowrap"><code><b>-v</b></code></td>
    <td class="t2">Print statistics about recorded packets on
      <code>stdout</code> while <i>rpld</i> is running. This option overrides
      <code>-s</code>.</td>
  </tr>
</table>

<h1>Privilege separation</h1>

<p class="block"><i>rpld</i> offers the possibility to change to another user's
identity after the initialization phase is complete, even though if it is not
really required since there is very few to exploit (if at all!), and on top
it is local. The default package (from <i>ttyrpld.sf.net</i>) uses the
<code>daemon</code> user. You can change this in the configuration file.</p>

<h1>Logging</h1>

<p class="block"><i>rpld</i> does not detach itself to help debugging, but
using the <code>startproc</code>, <code>setsid</code> starter tools or bash's
<code>disown</code> builtin can help bringing it in the background, if you need
to. The only time when the daemon will output something is either during
initialization or when there is really trouble, like memory allocation failure.
In either case, error messages are rare and you would not need to worry about
redirecting <code>stderr</code>.</p>

<p class="block">You can run <i>rpld</i> with the <code>-v</code> option to
enable printing statistics on <code>stdout</code>. Even if you do not have
<code>-v</code> specified, you can send <i>rpld</i> a <code>SIGALRM</code>
signal to make it print the current statistics. That of course only makes sense
when <code>stdout</code> is connected to something but <code>/dev/null</code>
or <code>/dev/zero</code>. You can send it multiple <code>SIGALRM</code>s, of
course, but you should consider using <code>-v</code> then, maybe. The
statistics are printed with the move-to-beginning-of-line character
(<code>\r</code>), which is not that suitable for logging, though.</p>

<p class="block">Basically, every tty is monitored, but certain kinds are
excluded, like the master sides of BSD (major number 2 (Linux), 6 on BSD) and
Unix98 ptys (major number 128), because they are just a mirror of their slave
sides with things turned around and are rarely useful. (On BSD, they are not
explicitly excluded because they do not seem to generate any data.)</p>

<p class="block"><i>rpld</i> will divert logging of a tty to a new file when
the tty inode is opened the next time and if the owner of it has changed. This
will make logins on <code>/dev/tty</code><i>N</i> go to the right file. Note
that <code>su</code> does not change the ownership, so both sessions (normal
and <code>su</code>'ed) will go into the same file.  (And that's good, because
it then logs by "real person" rather than login name.) The byte count
statistics are zeroed upon change detection.</p>

<h1>Configuration file</h1>

<p class="block"><i>rpld</i> starts with its hardcoded defaults, then tries to
read <code>/etc/rpld.conf</code> and finally <code>rpld.conf</code> in the
directory where the <code>rpld</code> binary is in.</p>

<p class="block">Configuration files have a simple <code>KEY=VALUE</code>
syntax. Empty lines, lines beginning with a hash mark (<code>#</code>) or
unrecognized keys are ignored.</p>

<table border="1">
  <tr>
    <td class="t1"><code>DEVICE</code></td>
    <td class="t1">List of possible rpl devices, separated by space. The
      default is "<code>/dev/misc/rpl /dev/rpl</code>".</td>
  </tr>
  <tr>
    <td class="t2"><code>OFMT</code></code>

    <td class="t2"><p class="block"><code>OFMT</code> combines the directory to
    write the tty recordings into (relative to the working directory of
    <i>rpld</i>) and the name of the log files. Note that the working directory
    of <i>rpld</i> depends on where it was started from, e.g. with
    <code>startproc</code>, the CWD is mostly the root path (<code>/</code>). 
    It is wise to use absolute paths here. The following printf-style tags can
    be used in <code>OFMT</code>:</p>

    <p class="block"><code>%d</code>&nbsp;-- Time when tty was opened (usually
    when someone logs in or an xterm was opened). In <code>YYYYMMDD</code>
    format.</p>

    <p class="block"><code>%l</code>&nbsp;-- Terminal (line) the user logged
    on.  The string is taken from a string database. Typical strings are
    <code>tty*</code> for virtual consoles, <code>pts-*</code> for
    pseudo-terminals, <code>ttyS*</code> for serial lines. If there is no
    matching string entry, the device number is used in square brackets, e.g.
    <code>[240:0]</code>.</p>

    <p class="block"><code>%u</code>&nbsp;-- User who owned the tty when it was
    opened. If the device could not be <code>stat()</code>'ed, <code>%u</code>
    will be substituted by "<code>NONE</code>". If the UID could not be
    translated into a name, the UID is spit out instead.</p>

    <p class="block"><i>Directories will be created as needed</i> (if
    permission allows). Thus, you can have
    <code>/var/log/rpl/%u/%d.%t.%l</code>, and it will automatically create
    <code>%u</code>, provided that <code>/var/log/rpl</code> is writable for
    the user running <i>rpld</i>.</p></td>

  </tr>
  <tr>
    <td class="t1"><code>USER</code></p>
    <td class="t1">User to change to after all initialization. Make sure that
      the user can create files according to <code>OFMT</code>.</td>
  </tr>
  <tr>
    <td class="t2"><code>START_INFOD</code></p>
    <td class="t2">Starts the <i>infod</i> component.</td>
  </tr>
  <tr>
    <td class="t1"><code>INFOD_PORT</code></p>
    <td class="t1">Specify the path of the socket which <i>infod</i> provides
      for clients. The default value is
      <code>/var/run/.rplinfod_socket</code>.</td>
  </tr>
</table>

<h1>See also</h1>

<ul>
  <li><a href="ttyreplay.1.php"><code>ttyreplay(1)</code></a></li>
</ul>

<?php include_once("_footer.php"); ?>
